store/tikv/lock_resolver.go

```diff
@@ -578,10 +598,249 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+	checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+		Keys:         curKeys,
+		StartVersion: txnID,
+	}
+	req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+	tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+	resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if regionErr != nil {
+		err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		logutil.BgLogger().Debug("checkSecondaries: region error, regrouping", zap.Uint64("txn id", txnID), zap.Uint64("region", curRegionID.GetID()))
+
+		// If regions have changed, then we might need to regroup the keys. Since this should be rare and for the sake
+		// of simplicity, we will resolve regions sequentially.
+		regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, curKeys, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for regionID, keys := range regions {
+			// Recursion will terminate because the resolve request succeeds or the Backoffer reaches its limit.
+			if err = lr.checkSecondaries(bo, txnID, keys, regionID, shared); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	if resp.Resp == nil {
+		return errors.Trace(ErrBodyMissing)
+	}
+
+	checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+	return shared.addKeys(checkResp.Locks, len(curKeys), txnID, checkResp.CommitTs)
+}
+
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus) error {
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.primaryLock.Secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	shared := asyncResolveData{
+		mutex:       sync.Mutex{},
+		commitTs:    status.primaryLock.MinCommitTs,
+		keys:        [][]byte{},
+		missingLock: false,
+	}
+
+	errChan := make(chan error, len(regions))
+
+	for regionID, keys := range regions {
+		curRegionID := regionID
+		curKeys := keys
+
+		go func() {
+			errChan <- lr.checkSecondaries(bo, l.TxnID, curKeys, curRegionID, &shared)
+		}()
+	}
+
+	var errs []string
+	for range regions {
+		err1 := <-errChan
+		if err1 != nil {
+			errs = append(errs, err1.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Errorf("async commit recovery (sending CheckSecondaryLocks) finished with errors: %v", errs)
+	}
+
+	shared.keys = append(shared.keys, l.Primary)
+	keysByRegion, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, shared.keys, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO(nrc, cfzjywxk) schema lease check
+
+	status.commitTS = shared.commitTs
+
+	errChan = make(chan error, len(shared.keys))
```

Comment:
I think the length should be `len(keysByRegion)`.

Reply:
good catch!